cpp/src/join/join_common_utils.cuh

```diff
@@ -0,0 +1,257 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <join/join_common_utils.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/sequence.h>
+#include <cub/cub.cuh>
+
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Computes the trivial left join operation for the case when the
+ * right table is empty. In this case all the valid indices of the left table
+ * are returned with their corresponding right indices being set to
+ * JoinNoneValue, i.e. -1.
+ *
+ * @param left Table of left columns to join
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the result
+ *
+ * @return Join output indices vector pair
+ */
+inline std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+                 std::unique_ptr<rmm::device_uvector<size_type>>>
+get_trivial_left_join_indices(
+  table_view const& left,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  auto left_indices = std::make_unique<rmm::device_uvector<size_type>>(left.num_rows(), stream, mr);
+  thrust::sequence(rmm::exec_policy(stream), left_indices->begin(), left_indices->end(), 0);
+  auto right_indices =
+    std::make_unique<rmm::device_uvector<size_type>>(left.num_rows(), stream, mr);
+  thrust::fill(
+    rmm::exec_policy(stream), right_indices->begin(), right_indices->end(), JoinNoneValue);
+  return std::make_pair(std::move(left_indices), std::move(right_indices));
+}
+
+// Convenient alias for a pair of unique pointers to device uvectors.
+using VectorPair = std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+                             std::unique_ptr<rmm::device_uvector<size_type>>>;
+
+/**
+ * @brief Takes two pairs of vectors and returns a single pair where the first
+ * element is a vector made from concatenating the first elements of both input
+ * pairs and the second element is a vector made from concatenating the second
+ * elements of both input pairs.
+ *
+ * This function's primary use is for computing the indices of a full join by
+ * first performing a left join, then separately getting the complementary
+ * right join indices, then finally calling this function to concatenate the
+ * results. In this case, each input VectorPair contains the left and right
+ * indices from a join.
+ *
+ * @param a The first pair of vectors.
+ * @param b The second pair of vectors.
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ *
+ * @return New pair of vectors containing the concatenated output.
```

Comment:
This isn't exactly true, is it? It looks like this function is mutating a and returning that or b in the case a is empty. Considering these are references coming in, we're invalidating the vector being passed in to the function. This seems a subtle side-effect that is either unintended or should be documented. I would lean towards mentioning it in the docs as the performance benefits of doing it like this can be large.

Reply:
Good point, yes we don't want to change this behavior for performance reasons but I will update the documentation to indicate that it mutates the inputs.